File: sdk/typescript/src/namespaces/explainability.ts
/**
 * Explainability Namespace API
 *
 * Provides a namespaced interface for decision explainability.
 * Essential for understanding AI decisions and building trust.
 */

import type {
  ExplainabilityResult,
  ExplanationFactor,
  CounterfactualScenario,
  CounterfactualList,
  CounterfactualGeneration,
  Narrative,
  Provenance,
  DebateEvidence,
} from '../types';

/**
 * Interface for the internal client methods used by ExplainabilityAPI.
 */
export type EvidenceResponse = DebateEvidence;

/**
 * Vote pivot item from a debate.
 */
export interface VotePivot {
  agent_id: string;
  vote: string;
  influence_score: number;
  reasoning?: string;
}

/**
 * Vote pivots response from getVotePivots.
 */
export interface VotePivotsResponse {
  debate_id: string;
  total_votes: number;
  pivotal_votes: number;
  pivot_threshold: number;
  votes: VotePivot[];
}

/**
 * Comparison result from compare.
 */
export interface ComparisonResponse {
  debates: Record<string, unknown>;
  comparison: {
    fields: Record<string, Array<{ debate_id: string; value: unknown }>>;
    summary: string;
  };
}

interface ExplainabilityClientInterface {
  getExplanation(debateId: string, options?: {
    include_factors?: boolean;
    include_counterfactuals?: boolean;
    include_provenance?: boolean;
  }): Promise<ExplainabilityResult>;
  getExplanationFactors(debateId: string, options?: {
    min_contribution?: number;
  }): Promise<{ factors: ExplanationFactor[] }>;
  getCounterfactuals(debateId: string, options?: {
    max_scenarios?: number;
  }): Promise<CounterfactualList>;
  generateCounterfactual(debateId: string, body: {
    hypothesis: string;
    affected_agents?: string[];
  }): Promise<CounterfactualGeneration>;
  getProvenance(debateId: string): Promise<Provenance>;
  getNarrative(debateId: string, options?: {
    format?: 'brief' | 'detailed' | 'executive_summary';
  }): Promise<Narrative>;
  createBatchExplanation(body: {
    debate_ids: string[];
    include_factors?: boolean;
    include_counterfactuals?: boolean;
    include_provenance?: boolean;
    include_narrative?: boolean;
  }): Promise<{ batch_id: string; status_url: string }>;
  getBatchExplanationStatus(batchId: string): Promise<{
    batch_id: string;
    status: 'pending' | 'processing' | 'completed' | 'failed';
    progress_pct: number;
  }>;
  getBatchExplanationResults(batchId: string, params?: {
    limit?: number;
    offset?: number;
  }): Promise<{ results: Array<{ debate_id: string; status: string; explanation?: unknown }> }>;
  getDebateEvidenceV1(debateId: string, options?: {
    limit?: number;
    min_relevance?: number;
  }): Promise<EvidenceResponse>;
  getDebateVotePivots(debateId: string, options?: {
    min_influence?: number;
  }): Promise<VotePivotsResponse>;
  compareExplanations(body: {
    debate_ids: string[];
    compare_fields?: string[];
  }): Promise<ComparisonResponse>;
}

/**
 * Explainability API namespace.
 *
 * Provides methods for understanding AI decisions:
 * - Get decision factors and their weights
 * - Generate counterfactual scenarios ("what if")
 * - Trace decision provenance
 * - Get human-readable narratives
 * - Batch process multiple debates
 *
 * Essential for building trust, compliance, and informed decision-making.
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Get full explanation for a debate
 * const explanation = await client.explainability.get('debate-123', {
 *   include_factors: true,
 *   include_counterfactuals: true
 * });
 *
 * // Get just the key factors
 * const { factors } = await client.explainability.getFactors('debate-123', {
 *   min_weight: 0.1
 * });
 *
 * // Get executive-friendly narrative
 * const narrative = await client.explainability.getNarrative('debate-123', {
 *   audience: 'executive'
 * });
 * console.log(narrative);
 *
 * // Generate "what if" scenario
 * const counterfactual = await client.explainability.generateCounterfactual('debate-123', {
 *   changes: ['Remove security agent', 'Add compliance requirement']
 * });
 * ```
 */
export class ExplainabilityAPI {
  constructor(private client: ExplainabilityClientInterface) {}

  // ===========================================================================
  // Core Explainability
  // ===========================================================================

  /**
   * Get full explanation for a debate decision.
   *
   * @param debateId - The debate ID to explain
   * @param options - What to include in the explanation
   */
  async get(debateId: string, options?: {
    include_factors?: boolean;
    include_counterfactuals?: boolean;
    include_provenance?: boolean;
  }): Promise<ExplainabilityResult> {
    return this.client.getExplanation(debateId, options);
  }

  /**
   * Get the factors that influenced a decision.
   *
   * Factors represent the key considerations that affected the outcome,
   * with contributions indicating their relative importance.
   *
   * @example
   * ```typescript
   * const { factors } = await client.explainability.getFactors('debate-123');
   * for (const factor of factors) {
   *   console.log(`${factor.name}: ${factor.contribution > 0 ? '+' : ''}${factor.contribution}`);
   * }
   * ```
   */
  async getFactors(debateId: string, options?: {
    min_contribution?: number;
  }): Promise<{ factors: ExplanationFactor[] }> {
    return this.client.getExplanationFactors(debateId, options);
  }

  // ===========================================================================
  // Counterfactuals
  // ===========================================================================

  /**
   * Get existing counterfactual scenarios for a debate.
   *
   * @param options.max_scenarios - Maximum number of scenarios to return
   */
  async getCounterfactuals(debateId: string, options?: {
    max_scenarios?: number;
  }): Promise<CounterfactualList> {
    return this.client.getCounterfactuals(debateId, options);
  }

  /**
   * Generate a new counterfactual scenario.
   *
   * Counterfactuals help understand how changes would affect the decision.
   *
   * @example
   * ```typescript
   * const counterfactual = await client.explainability.generateCounterfactual('debate-123', {
   *   hypothesis: 'What if we had a larger budget?',
   *   affected_agents: ['gpt-4', 'claude']
   * });
   * console.log('Predicted outcome:', counterfactual.predicted_outcome);
   * ```
   */
  async generateCounterfactual(debateId: string, body: {
    hypothesis: string;
    affected_agents?: string[];
  }): Promise<CounterfactualGeneration> {
    return this.client.generateCounterfactual(debateId, body);
  }

  // ===========================================================================
  // Provenance
  // ===========================================================================

  /**
   * Get the provenance (source tracking) for a decision.
   *
   * Traces where information came from and how reasoning progressed.
   */
  async getProvenance(debateId: string): Promise<Provenance> {
    return this.client.getProvenance(debateId);
  }

  // ===========================================================================
  // Narrative
  // ===========================================================================

  /**
   * Get a human-readable narrative explanation.
   *
   * @param debateId - The debate ID
   * @param options.format - Format of the narrative:
   *   - brief: Short summary
   *   - detailed: Full explanation
   *   - executive_summary: High-level executive summary
   */
  async getNarrative(debateId: string, options?: {
    format?: 'brief' | 'detailed' | 'executive_summary';
  }): Promise<Narrative> {
    return this.client.getNarrative(debateId, options);
  }

  // ===========================================================================
  // Batch Processing
  // ===========================================================================

  /**
   * Create a batch explanation request for multiple debates.
   *
   * Useful for generating reports across many decisions.
   *
   * @example
   * ```typescript
   * const { batch_id } = await client.explainability.createBatch({
   *   debate_ids: ['debate-1', 'debate-2', 'debate-3'],
   *   include_factors: true,
   *   include_narrative: true
   * });
   *
   * // Poll for completion
   * let status = await client.explainability.getBatchStatus(batch_id);
   * while (status.status === 'processing') {
   *   await sleep(1000);
   *   status = await client.explainability.getBatchStatus(batch_id);
   * }
   *
   * // Get results
   * const { results } = await client.explainability.getBatchResults(batch_id);
   * ```
   */
  async createBatch(request: {
    debate_ids: string[];
    include_factors?: boolean;
    include_counterfactuals?: boolean;
    include_provenance?: boolean;
    include_narrative?: boolean;
  }): Promise<{ batch_id: string; status_url: string }> {
    return this.client.createBatchExplanation(request);
  }

  /**
   * Get the status of a batch explanation request.
   */
  async getBatchStatus(batchId: string): Promise<{
    batch_id: string;
    status: 'pending' | 'processing' | 'completed' | 'failed';
    progress_pct: number;
  }> {
    return this.client.getBatchExplanationStatus(batchId);
  }

  /**
   * Get the results of a completed batch explanation.
   */
  async getBatchResults(batchId: string, params?: {
    limit?: number;
    offset?: number;
  }): Promise<{ results: Array<{ debate_id: string; status: string; explanation?: unknown }> }> {
    return this.client.getBatchExplanationResults(batchId, params);
  }

  // ===========================================================================
  // Evidence
  // ===========================================================================

  /**
   * Get the evidence chain for a debate decision.
   *
   * Returns all evidence and sources cited during the debate,
   * sorted by relevance.
   *
   * @param debateId - The debate ID
   * @param options.limit - Maximum number of evidence items to return
   * @param options.min_relevance - Minimum relevance score (0-1)
   *
   * @example
   * ```typescript
   * const evidence = await client.explainability.getEvidence('debate-123', {
   *   limit: 10,
   *   min_relevance: 0.5
   * });
   * console.log(`Found ${evidence.evidence_count} evidence items`);
   * console.log(`Quality score: ${evidence.evidence_quality_score}`);
   * ```
   */
  async getEvidence(debateId: string, options?: {
    limit?: number;
    min_relevance?: number;
  }): Promise<EvidenceResponse> {
    return this.client.getDebateEvidenceV1(debateId, options);
  }

  // ===========================================================================
  // Vote Analysis
  // ===========================================================================

  /**
   * Get vote pivots analysis for a debate.
   *
   * Identifies which agent votes were pivotal in determining the outcome,
   * showing influence scores and reasoning.
   *
   * @param debateId - The debate ID
   * @param options.min_influence - Minimum influence score to include
   *
   * @example
   * ```typescript
   * const pivots = await client.explainability.getVotePivots('debate-123');
   * console.log(`${pivots.pivotal_votes} of ${pivots.total_votes} votes were pivotal`);
   * for (const vote of pivots.votes) {
   *   console.log(`${vote.agent_id}: ${vote.vote} (influence: ${vote.influence_score})`);
   * }
   * ```
   */
  async getVotePivots(debateId: string, options?: {
    min_influence?: number;
  }): Promise<VotePivotsResponse> {
    return this.client.getDebateVotePivots(debateId, options);
  }

  // ===========================================================================
  // Comparison
  // ===========================================================================

  /**
   * Compare explanations between multiple debates.
   *
   * Useful for understanding patterns across decisions or
   * identifying why different debates had different outcomes.
   *
   * @param body.debate_ids - IDs of debates to compare (2-10)
   * @param body.compare_fields - Fields to compare (default: confidence, consensus_reached, contributing_factors, evidence_quality)
   *
   * @example
   * ```typescript
   * const comparison = await client.explainability.compare({
   *   debate_ids: ['debate-1', 'debate-2', 'debate-3'],
   *   compare_fields: ['confidence', 'evidence_quality']
   * });
   * console.log('Summary:', comparison.comparison.summary);
   * ```
   */
  async compare(body: {
    debate_ids: string[];
    compare_fields?: string[];
  }): Promise<ComparisonResponse> {
    return this.client.compareExplanations(body);
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  /**
   * Get the top N positive factors from an explanation.
   */
  getTopPositiveFactors(result: ExplainabilityResult, n: number = 5): ExplanationFactor[] {
    return (result.factors ?? [])
      .filter(f => f.contribution > 0)
      .sort((a, b) => b.contribution - a.contribution)
      .slice(0, n);
  }

  /**
   * Get the top N negative factors from an explanation.
   */
  getTopNegativeFactors(result: ExplainabilityResult, n: number = 5): ExplanationFactor[] {
    return (result.factors ?? [])
      .filter(f => f.contribution < 0)
      .sort((a, b) => a.contribution - b.contribution)
      .slice(0, n);
  }
}

// Re-export types for convenience
// Re-export types from this namespace
export type { ExplainabilityResult, ExplanationFactor, CounterfactualScenario, Narrative, Provenance };

File: aragora-py/aragora_client/explainability.py
"""Explainability API for the Aragora SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

from pydantic import BaseModel

if TYPE_CHECKING:
    from aragora_client.client import AragoraClient


class ExplanationFactor(BaseModel):
    """Factor contributing to a decision."""

    name: str
    contribution: float
    description: str | None = None
    evidence: list[str] | None = None


class Explanation(BaseModel):
    """Decision explanation."""

    debate_id: str
    summary: str
    factors: list[ExplanationFactor] | None = None
    confidence: float | None = None
    methodology: str | None = None
    created_at: str | None = None


class Counterfactual(BaseModel):
    """Counterfactual scenario."""

    id: str
    scenario: str
    outcome: str
    probability: float | None = None
    key_changes: list[str] | None = None


class ProvenanceEntry(BaseModel):
    """Provenance tracking entry."""

    source: str
    timestamp: str
    action: str
    agent: str | None = None
    details: dict[str, Any] | None = None


class Narrative(BaseModel):
    """Decision narrative."""

    debate_id: str
    format: str
    content: str
    word_count: int | None = None
    created_at: str | None = None


class BatchExplanationJob(BaseModel):
    """Batch explanation job status."""

    batch_id: str
    status: str
    total: int
    completed: int
    failed: int
    created_at: str | None = None
    completed_at: str | None = None


class ExplainabilityAPI:
    """API for decision explainability operations."""

    def __init__(self, client: AragoraClient) -> None:
        self._client = client

    # ==========================================================================
    # Explanations
    # ==========================================================================

    async def get_explanation(
        self,
        debate_id: str,
        *,
        format: Literal["brief", "detailed", "technical"] | None = None,
        include_factors: bool = True,
    ) -> Explanation:
        """Get explanation for a debate decision.

        Args:
            debate_id: Debate ID
            format: Explanation format (brief, detailed, technical)
            include_factors: Whether to include contributing factors

        Returns:
            Decision explanation
        """
        params: dict[str, Any] = {"include_factors": include_factors}
        if format:
            params["format"] = format

        data = await self._client._get(
            f"/api/v1/debates/{debate_id}/explanation", params=params
        )
        return Explanation.model_validate(data)

    async def get_factors(
        self,
        debate_id: str,
        *,
        min_contribution: float | None = None,
    ) -> list[ExplanationFactor]:
        """Get factors contributing to a debate decision.

        Args:
            debate_id: Debate ID
            min_contribution: Minimum contribution threshold (0-1)

        Returns:
            List of contributing factors
        """
        params: dict[str, Any] = {}
        if min_contribution is not None:
            params["min_contribution"] = min_contribution

        data = await self._client._get(
            f"/api/v1/debates/{debate_id}/explanation/factors",
            params=params or None,
        )
        return [ExplanationFactor.model_validate(f) for f in data.get("factors", [])]

    # ==========================================================================
    # Counterfactuals
    # ==========================================================================

    async def get_counterfactuals(
        self,
        debate_id: str,
        *,
        max_scenarios: int | None = None,
    ) -> list[Counterfactual]:
        """Get counterfactual scenarios for a debate.

        Args:
            debate_id: Debate ID
            max_scenarios: Maximum number of scenarios to generate

        Returns:
            List of counterfactual scenarios
        """
        params: dict[str, Any] = {}
        if max_scenarios is not None:
            params["max_scenarios"] = max_scenarios

        data = await self._client._get(
            f"/api/v1/debates/{debate_id}/counterfactuals",
            params=params or None,
        )
        return [
            Counterfactual.model_validate(c) for c in data.get("counterfactuals", [])
        ]

    async def generate_counterfactual(
        self,
        debate_id: str,
        *,
        variable: str,
        new_value: str,
        context: str | None = None,
    ) -> Counterfactual:
        """Generate a specific counterfactual scenario.

        Args:
            debate_id: Debate ID
            variable: Variable to change
            new_value: New value for the variable
            context: Additional context for generation

        Returns:
            Generated counterfactual scenario
        """
        body: dict[str, Any] = {
            "variable": variable,
            "new_value": new_value,
        }
        if context:
            body["context"] = context

        data = await self._client._post(
            f"/api/v1/debates/{debate_id}/counterfactuals/generate", body
        )
        return Counterfactual.model_validate(data)

    # ==========================================================================
    # Provenance
    # ==========================================================================

    async def get_provenance(
        self,
        debate_id: str,
        *,
        limit: int = 50,
        offset: int = 0,
    ) -> list[ProvenanceEntry]:
        """Get provenance chain for a debate.

        Args:
            debate_id: Debate ID
            limit: Maximum number of entries
            offset: Pagination offset

        Returns:
            List of provenance entries
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        data = await self._client._get(
            f"/api/v1/debates/{debate_id}/provenance", params=params
        )
        return [ProvenanceEntry.model_validate(p) for p in data.get("entries", [])]

    # ==========================================================================
    # Narratives
    # ==========================================================================

    async def get_narrative(
        self,
        debate_id: str,
        *,
        format: Literal["brief", "detailed", "executive_summary"] = "detailed",
    ) -> Narrative:
        """Get narrative explanation for a debate.

        Args:
            debate_id: Debate ID
            format: Narrative format

        Returns:
            Decision narrative
        """
        params: dict[str, Any] = {"format": format}
        data = await self._client._get(
            f"/api/v1/debates/{debate_id}/narrative", params=params
        )
        return Narrative.model_validate(data)

    # ==========================================================================
    # Batch Operations
    # ==========================================================================

    async def create_batch_explanation(
        self,
        debate_ids: list[str],
        *,
        format: Literal["brief", "detailed", "technical"] = "detailed",
        include_factors: bool = True,
        include_counterfactuals: bool = False,
    ) -> BatchExplanationJob:
        """Create batch explanation job for multiple debates.

        Args:
            debate_ids: List of debate IDs
            format: Explanation format
            include_factors: Include contributing factors
            include_counterfactuals: Include counterfactual scenarios

        Returns:
            Batch job status
        """
        body: dict[str, Any] = {
            "debate_ids": debate_ids,
            "format": format,
            "include_factors": include_factors,
            "include_counterfactuals": include_counterfactuals,
        }
        data = await self._client._post("/api/v1/explanations/batch", body)
        return BatchExplanationJob.model_validate(data)

    async def get_batch_status(self, batch_id: str) -> BatchExplanationJob:
        """Get status of a batch explanation job.

        Args:
            batch_id: Batch job ID

        Returns:
            Batch job status
        """
        data = await self._client._get(f"/api/v1/explanations/batch/{batch_id}")
        return BatchExplanationJob.model_validate(data)

    async def get_batch_results(
        self,
        batch_id: str,
        *,
        limit: int = 50,
        offset: int = 0,
    ) -> list[Explanation]:
        """Get results from a batch explanation job.

        Args:
            batch_id: Batch job ID
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            List of explanations
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        data = await self._client._get(
            f"/api/v1/explanations/batch/{batch_id}/results", params=params
        )
        return [Explanation.model_validate(e) for e in data.get("results", [])]
